File: cpp/src/handlers/file.cpp
/**
 * @file file.cpp
 * @brief File handler implementation
 */

#include <agora/log/handlers/file.hpp>
#include <agora/log/formatter.hpp>
#include <stdexcept>

namespace agora::log {

FileHandler::FileHandler(const std::filesystem::path& file_path)
    : file_path_(file_path) {
    open_file();
}

FileHandler::~FileHandler() noexcept {
    close_file();
}

void FileHandler::write(const LogEntry& entry) {
    std::string formatted = format_json(entry);

    std::lock_guard<std::mutex> lock(mutex_);

    if (!file_.is_open()) {
        open_file();
    }

    file_ << formatted << '\n';
}

void FileHandler::flush() noexcept {
    try {
        std::lock_guard<std::mutex> lock(mutex_);
        if (file_.is_open()) {
            file_.flush();
        }
    } catch (...) {
        // Ignore errors during flush - noexcept guarantee
    }
}

void FileHandler::open_file() {
    // Create parent directories if they don't exist
    if (file_path_.has_parent_path()) {
        std::filesystem::create_directories(file_path_.parent_path());
    }

    file_.open(file_path_, std::ios::app);

    if (!file_.is_open()) {
        throw std::runtime_error("Failed to open log file: " + file_path_.string());
    }
}

void FileHandler::close_file() noexcept {
    try {
        if (file_.is_open()) {
            file_.flush();
            file_.close();
        }
    } catch (...) {
        // Ignore errors during close - noexcept guarantee
    }
}

}  // namespace agora::log

File: python/src/agora_log/handlers/file.py
"""File handler for logging output."""

import threading
from pathlib import Path
from typing import Any

from .base import Handler
from ..formatter import format_json


class FileHandler(Handler):
    """
    Handler that writes log entries to a file.

    Thread-safe file writes with automatic directory creation.
    Always uses JSON format for file output.
    """

    def __init__(self, file_path: Path):
        """
        Initialize file handler.

        Args:
            file_path: Path to the log file
        """
        self.file_path = Path(file_path)
        self._lock = threading.Lock()
        self._file = None

        # Create parent directories if they don't exist
        self.file_path.parent.mkdir(parents=True, exist_ok=True)

        # Open file in append mode
        self._open()

    def _open(self) -> None:
        """Open the log file for writing."""
        self._file = open(self.file_path, "a", encoding="utf-8")

    def emit(self, entry: dict[str, Any]) -> None:
        """Write log entry to file."""
        try:
            output = format_json(entry)
            with self._lock:
                if self._file and not self._file.closed:
                    self._file.write(output + "\n")
                    self._file.flush()
        except Exception:
            # Fail silently - logging should not crash the app
            pass

    def flush(self) -> None:
        """Flush the file buffer."""
        try:
            with self._lock:
                if self._file and not self._file.closed:
                    self._file.flush()
        except Exception:
            pass

    def close(self) -> None:
        """Close the file handler."""
        try:
            with self._lock:
                if self._file and not self._file.closed:
                    self._file.flush()
                    self._file.close()
        except Exception:
            pass
